camel/bots/discord/discord_app.py

```
@@ -0,0 +1,375 @@
+                logger.error(
+                    f"Failed to refresh access token for guild: {guild_id}"
+                )
+                return None
+
+        return installation.access_token
+
+    async def save_installation(
+        self,
+        guild_id: str,
+        access_token: str,
+        refresh_token: str,
+        expires_in: int,
+    ):
+        r"""Save the installation information for a given guild.
+
+        Args:
+            guild_id (str): The ID of the guild where the bot is installed.
+            access_token (str): The access token for the guild.
+            refresh_token (str): The refresh token for the guild.
+            expires_in: (int): The expiration time of the
+                access token.
+        """
+        if not self.oauth_flow:
+            logger.warning(
+                "OAuth is not enabled. Missing client_id, "
+                "client_secret, or redirect_uri."
+            )
+            return None
+        assert self.installation_store is not None
+        expires_at = datetime.now() + timedelta(seconds=expires_in)
+        installation = DiscordInstallation(
+            guild_id=guild_id,
+            access_token=access_token,
+            refresh_token=refresh_token,
+            installed_at=datetime.now(),
+            token_expires_at=expires_at,
+        )
+        await self.installation_store.save(installation)
+        logger.info(f"Installation saved for guild: {guild_id}")
+
+    async def remove_installation(self, guild: discord.Guild):
+        r"""Remove the installation for a given guild.
+
+        Args:
+            guild (discord.Guild): The guild from which the bot is
+                being removed.
+        """
+        if not self.oauth_flow:
+            logger.warning(
+                "OAuth is not enabled. Missing client_id, "
+                "client_secret, or redirect_uri."
+            )
+            return None
+        assert self.installation_store is not None
```

It would be better to raise an error here.